### ISSUE TITLE
Figure out what Kubernetes installation method we 'support'

### ISSUE THREAD

**Issue by yuvipanda** [jupyterhub/zero-to-jupyterhub-k8s#593]
We currently have instructions for Google, Azure and AWS to create a kubernetes cluster. We've had people try to set up clusters in other places:

1. JetStream (from @zonca https://zonca.github.io/2017/12/scalable-jupyterhub-kubernetes-jetstream.html)
2. Digital Ocean (https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/401)
3. Bare metal clusters in various places (https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/591, https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/277, https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/493 and various questions on the Gitter Channel)
4. OpenStack (https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/88, https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/340, https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/237#issuecomment-341523196 and various discussions)
5. OpenShift

I'd love for us to answer the following questions:

1. When do we have instructions in the guide, and when do we just link to an external location?
2. What do we do when users have problems with kubernetes installations and come to us? Do we help try to fix them? Do we redirect them to the k8s community? What is the line here between 'kubernetes problem' and 'JupyterHub' problem?
3. For the providers we have instructions for, should we grade them with 'levels of support'? Which ones do we test with Kubernetes version updates and JupyterHub chart version updates?

I think we need to make sure we communicate clearly that you can use z2jh without depending on Cloud Providers. This is especially important in the current climate of intense critique of privacy of various platforms. Several people have run k8s clusters successfully on all these platforms, but everyone has to re-learn the same lessons without any place to share them.

Here are the various options I see:

1. Introduce various installation methods in the z2jh guide that can be contritbuted to by users. We would answer question (3) from above, and mark these with appropriate levels. Particularly important is how much testing we do with new releases, since we might not have access to test on these platforms.
2. Have a place to put links to various blog posts. This takes all decisions out of our hands, and 'reader beware, check the date before trying!' mode.
3. Have a wiki or a community repo for doing option (1), to make it clearer that it is not part of the main z2jh guide. This can have much wider merge rights - perhaps anyone who lands a PR gets merge rights? 

These are not exclusive, and there are definitely other options! Just writing these here to kick off conversation :)

**Comment by yuvipanda** [comment]
Tagging @gsemet who was interested in doing this for OpenStack I believe.

Tagging @ZachGlassman who was interested in doing this for DigitalOcean.

Tagging @aculich who was interested in seeing this happen for bare metal installations (but might not have time himself)

Tagging @choldgraf who told me that this was #1 asked for feature in his recently concluded tour of Europe :)

**Comment by choldgraf** [comment]
Just adding a quick bit of info: the "#1 asked for feature" was basically "I want to run JupyterHub on k8s without giving google/amazon/MS/<some big company> a bunch of money"

**Comment by manics** [comment]
I'm happy to help with OpenStack too. However maybe this could be combined with a bare-metal installation, especially if the instructions are designed for a "minimal functionality" Kubernetes cluster:
- no load-balancer:
  - Assume either an ingress controller or if that's not configured then nodeports
- no dynamic storage:
  - Maybe assume a single NFS share and use chown as in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/421
  - Or use something from https://github.com/kubernetes-incubator/external-storage, I've had success with the [dynamic NFS provisioner](https://github.com/kubernetes-incubator/external-storage/tree/233b44b89dee15d9ec4ba2e084abb74735a60909/nfs)

**Comment by gsemet** [comment]
Hi. Thanks for this notification. We don't use openstack but a custom baremetal Kube cluster (based on CoreOS). I did not have any major change on the guide that was an excellent support for my setup and self education.

Jupyterhub works like a charm behind a traefik ingress, I'll be happy to document on this use case + weird/experimental helming tricks (postgres, ...).

We do have weird restrictions imposed by the team in charge of the cluster, for instance all non officially supported software should be in the same namespace, no DBaaS, and so on, so I usually open new tickets here :)

But I would be glad to contribute to a community wiki! I do not recommend the blog section (a blog section in the wiki is perfectly acceptable). Blog links breaks, can be no more maintained by the author, a wiki is the perfect tool to ensure a useful information stays up to date even in case of maintenance rotation.

**Comment by choldgraf** [comment]
I am also +1 on number 3 and _probably_ number 1, but -1 on linking to blog posts for the reasons others have stated (e.g. we don't want to create a project norm of "oh don't use that blog post, use _this_ newer blog post")

**Comment by ZachGlassman** [comment]
Hi.  I would be happy to contribute to a community wiki.  Like above , I have been using https://github.com/kubernetes-incubator/external-storage with success for a while and could probably provide some wiring tips on that.

**Comment by zonca** [comment]
I think a wiki or community repo would be fine, and it is important to publish guidelines that set how much freedom a user has to edit an existing recipe. Or, do we need a group of maintainers for each recipe (swcarpentry style)?

**Comment by yuvipanda** [comment]
Also /cc @willingc and @jzf2101, who have probably put a lot of thought into structuring documentation.

**Comment by choldgraf** [comment]
totally agree regarding the "dividing line" -  I think that there one big difference that separates out two "types" of vendors:  one group that offers some kind of "click this button and we'll give you a k8s cluster" service, and the group where you have to set all of that up yourself.

My intuition is that the former group is much easier to "officially" support/maintain within z2jh, and the latter will require some other process to generate community knowledge on bootstrapping your own k8s cluster.

Either way, I like the idea of modularizing the vendor-specific docs rather than just glomming them on to the one page :-)

**Comment by willingc** [comment]
FYI. @choldgraf I'm going to break out the "Creating a Kubernetes cluster" on the main docs page into its own section from "Creating your JupyterHub" section. Working on a WIP PR that we can iterate on or pitch if we come up with something better.

**Comment by choldgraf** [comment]
Cool! Excited to see how the PR looks :-) thanks for taking an initial stab!

**Comment by cam72cam** [comment]
I'd like to do a AWS kops, but without using Heptio.  Would that be welcome?

**Comment by yuvipanda** [comment]
@cam72cam most definitely! #243 is the issue for that.

**Comment by willingc** [comment]
I think that PR #594 gives us a good base to work off for adding community docs.

**Comment by minrk** [comment]
I think for the most part, z2jh should not have docs on setting up kubernetes. Copy/paste commands for cloud providers that give you easy, fully-functional clusters makes sense, but openstack/kops/kubeadm I think ought to be out of scope. For these, I think we should link to *external* docs on setting up kubernetes clusters, but I don't want to put us in a position to be getting bug reports about deploying a kubernetes cluster since there is an endless supply of options and things that can go wrong. But some description of what is required of a kubernetes setup that we require is appropriate, so that people looking to deploy their own kubernetes can know what they need to have. I think the main points for supporting less-than-complete kubernetes installations is the dynamic storage and load-balancer, which are often absent from self-deployed clusters, as mentioned above. It *may* be appropriate to include "if you don't have a load-balancer" and/or "if you don't have a storage provider" configuration examples.

**Comment by gsemet** [comment]
I agree. There are many configuration of kube cluster, the setup of these config should not, IMHO, be under the umbrella of z2jh, but you can give hints on what to do given the users configuration. Maybe a full example using minikube (with an ingress) can help newbies, I have harder time using minikube now than using our cluster, now I know better what our cluster can do and cannot.
And if I move to another cluster (GCE, ...), I will be happy to find these different options described in z2jh guide.

**Comment by gsemet** [comment]
sounds great! Makes sense. Do you think you (z2jh) could still host a wiki (mediawiki, github wiki or any other) for the community to self-organise around?

**Comment by yuvipanda** [comment]
@gsemet I'd like to, but let's open a different issue specific to discussing that? I think these are all big questions, and keeping discussion focused is important.

**Comment by jzf2101** [comment]
@consideRatio Also this as reference https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/646

**Comment by willingc** [comment]
I should be getting early access to the Digital Ocean service when I return home.

**Comment by alexmorley** [comment]
Hey all. From what I've experience so far the Digital Ocean Kubernetes (still in beta) offering is easier to get start on than anything else I've tried so far (Azure and GCP). I've started a PR with instructions to set it up at #1192 .

**Comment by ablekh** [comment]
@alexmorley Nice! Just keep in mind that AFAIK current feature set of Digital Ocean's managed K8s offering is not fully comparable with major public cloud providers due to lack of automatic upgrades, auto-scaling, multiple availability zones, integration with AD etc. It would be also interesting to compare pricing with usual suspects as well as performance (cluster creation/deletion times). Do you have any comments on this?

**Comment by consideRatio** [comment]
Oh thanks for the insights about this @ablekh!

**Comment by ablekh** [comment]
@consideRatio My pleasure! :-)

**Comment by betatim** [comment]
I think we should keep our guides as unopinionated as possible. If there are features missing that a basic JupyterHub needs then we shouldn't have a guide but if features beyond that are missing that somehow should be left as a judgement to the user. We shouldn't become a place that passes judgement on what k8s is "the best". We already struggle just to provide guides for setup, also having to make defendable and fair judgements on additional features would sidetrack us IMHO.

Not sure what the best place for sharing such additional info is though as it is clearly interesting and relevant for people wanting to pick one :-/

**Comment by alexmorley** [comment]
> We already struggle just to provide guides for setup, also having to make defendable and fair judgements on additional features would sidetrack us IMHO.

I agree. If possible finding an external resource for choosing a k8s provider and link out to that.

**Comment by atulyadavtech** [comment]
Add information about Bare metal clusters installation.  Looking forward...

**Comment by consideRatio** [comment]
I think we need some notes on bare metal clusters, as they lead to questions about "why is my LoadBalancer k8s service pending?" and such which we could address there, but overall I think the practical path we have taken so far have been that if someone does the big job of adding a big chunk on how to install z2jh on a certain cloud, then we have added it. Maintenance have been reasonable in my mind.

Closing this in favor of another issue now, where I summarize the wish for more on bare metal clusters.
